apis/v1alpha1/udproute_types.go

```diff
@@ -0,0 +1,109 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// UDPRouteSpec defines the desired state of UDPRoute
+type UDPRouteSpec struct {
+	// Rules are a list of UDP matchers and actions.
+	Rules []UDPRouteRule `json:"rules" protobuf:"bytes,1,rep,name=rules"`
+}
+
+// UDPRouteStatus defines the observed state of UDPRoute
+type UDPRouteStatus struct {
+	GatewayRefs []GatewayObjectReference `json:"gatewayRefs" protobuf:"bytes,1,rep,name=gatewayRefs"`
+}
+
+// +genclient
+// +kubebuilder:object:root=true
+
+// UDPRoute is the Schema for the udproutes API
+type UDPRoute struct {
+	metav1.TypeMeta   `json:",inline" protobuf:"bytes,1,opt,name=typeMeta"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,2,opt,name=metadata"`
+
+	Spec   UDPRouteSpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
+	Status UDPRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+}
+
+// UDPRouteRule is the configuration for a given rule.
+type UDPRouteRule struct {
+	// Match defines which packets match this rule.
+	//
+	// +optional
+	Match *UDPRouteMatch `json:"match" protobuf:"bytes,1,opt,name=match"`
+	// Action defines what happens to the packet.
+	//
+	// +optional
+	Action *UDPRouteAction `json:"action" protobuf:"bytes,2,opt,name=action"`
+}
+
+// UDPRouteAction is the action for a given rule.
+type UDPRouteAction struct {
+	// ForwardTo sends requests to the referenced object.  The
+	// resource may be "services" (omit or use the empty string for the
+	// group), or an implementation may support other resources (for
+	// example, resource "myroutetargets" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "services".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the UDPRoute will be true.
+	ForwardTo *ForwardToTarget `json:"forwardTo" protobuf:"bytes,1,opt,name=forwardTo"`
+
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "action" behavior.  The resource may be "configmaps" (use the empty
```

Comment:
This is the same language used in TCPRoute, so it's fine for this PR. However, as written it's not clear whether the extension *is* an action (i.e. does something) or merely *modifies* an action (i.e. extends ForwardTo in some capacity). Maybe we should clarify.